Extension point for languages

To have an awesome language agnostic experience it's nice if the sdk (like previous `preview2`) has a clean extension point to add additional languages

Atm works ( ref f# [examples](https://github.com/dotnet/netcorecli-fsc/tree/master/examples/preview3) and [tests suite](https://github.com/dotnet/netcorecli-fsc/tree/master/test-msbuild) ), but can be improved a bit

An example the [fsharp with preview3](https://github.com/dotnet/netcorecli-fsc/blob/99e22f5413fa0dc7ea16bf92a35fb0048fa99207/examples/preview3/console/ExampleConsoleApp.fsproj#L26-L28)

So:

1. Add a `PackageReference`  to a lang sdk package who contains msbuild target files (example [FSharp.NET.Sdk](https://github.com/dotnet/netcorecli-fsc/tree/master/src/FSharp.NET.Sdk) ) who are auto imported
     - The targets file contains a:
          - `CoreCompile` task => ok, thats usually invoke the compiler to compile
          - `CreateManifestResourceNames` task => why?
2. Import at the end `$(MSBuildToolsPath)\Microsoft.Common.targets` (instead of `Microsoft.CSharp.targets`)

That's ok and works, but with some issues:

- I cannot directly include `Microsoft.Common.targets` inside my target files, because at the first `dotnet restore` doesnt exists, so restore doesnt work (`Common` imports nuget restore target somewhere).
- I think is important to be able to import `Microsoft.Common.targets` inside custom language targets, for example to be able to replicate the .net framework build script (so wildcard include before and after `.Common` import). If i put `.Common` inside the target file is an authoring error because is included two times (on `dotnet restore` is ok, but after that is included two times)
- I can specify to import the target inside nugetpackage at the end of project file, but is annoying atm because i need to do that starting from $NugetPackageCacheRoot directory, so that's going to include version of package. Also doesnt work at first `dotnet restore`, because doesnt exists yet, so i need to conditionally include `Common` anyway

Best experience from project:

- Just include a `PackageReference` with custom targets like `CoreCompile` (compiler can be a dotnet sdk tool :love:)
- No need to include `Microsoft.Common.targets` target at the end, maybe another target who will invoke the auto-imported lang target.

@nguerrera you have some ideas about how to do the extensibility changes?

/cc @piotrpMSFT @livarcocc @blackdwarf 
/cc @cartermp @KevinRansom @cloudRoutine for F#


Moved from https://github.com/dotnet/cli/issues/4669 on behalf of @enricosada

@nguerrera @rainersigwald nuget.target alone doesnt work.

- I used https://github.com/dotnet/netcorecli-fsc/tree/master/examples/preview3/console
- imported only `<Import Project="$(MBuildExtensionsPath)\NuGet.targets" />` at end (no Common)
- `dotnet restore`, fails because `$(MBuildExtensionsPath)` was empty, so joined path was invalid path
- changed to `<Import Project="$(MSBuildToolsPath)\NuGet.targets" />` , another `dotnet restore`
- fails too, but now i get `target "_SplitProjectReferencesByFileExistence doesnt exists`

    ```
    c:\dotnetcli\dotnet-dev-win-x64.1.0.0-preview3-004056\sdk\1.0.0-preview3-004056\NuGet.targets(538,5): error MSB4057: The target "_SplitProjectReferencesByFileExistence" does not exist in the project. [e:\github\netcorecli-fsc\examples\preview3\console\ExampleConsoleApp.fsproj]
    ```

So yes, depends on some bit elsewhere 😄 .

@nguerrera @rainersigwald to try you can also use `https://github.com/dotnet/netcorecli-fsc/tree/master/examples/preview3/console` as example if you want. The downloaded f# sdk targets are inside the `FSharp.NET.Sdk` package,, so you can mess with local .fsproj and the msbuild targets in nuget cache (a bonus of sdk packages, remove it from cache fix mistakes next restore)


As a note: I think having a clean language extension point help everyone, c#, VB, f#.
In `preview2` the `dotnet-compile-csc`  (for c#) was embedded in sdk, but the extension point was `dotnet-compile-{compilerName}` so worked really well.
That can help also c# if they want to move sdk integration as packageref (has lots of benefits 😄 ), just sayin

--- a/src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.targets
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/sdk/Sdk.targets
@@ -28,10 +28,13 @@ Copyright (c) .NET Foundation. All rights reserved.
   <PropertyGroup Condition="'$(LanguageTargets)' == ''">
     <LanguageTargets Condition="'$(MSBuildProjectExtension)' == '.csproj'">$(MSBuildToolsPath)\Microsoft.CSharp.targets</LanguageTargets>
     <LanguageTargets Condition="'$(MSBuildProjectExtension)' == '.vbproj'">$(MSBuildToolsPath)\Microsoft.VisualBasic.targets</LanguageTargets>
+    
+    <!-- If LanguageTargets isn't otherwise set, then just import the common targets.  This should allow the restore target to run,
+         which could bring in NuGet packages that set the LanguageTargets to something else.  This means support for different
+         languages could either be supplied via an SDK or via a NuGet package. -->
+    <LanguageTargets Condition="'$(LanguageTargets)' == ''">$(MSBuildToolsPath)\Microsoft.Common.CurrentVersion.targets</LanguageTargets>
   </PropertyGroup>
-  
-  <!-- TODO: Generate error if LanguageTargets property isn't set here.  This would happen, for example if an .fsproj referenced the .NET Sdk 
-              but not the FSharp one.  See https://github.com/dotnet/sdk/issues/448 -->
+
   <!-- REMARK: Dont remove/rename, the LanguageTargets property is used by F# to hook inside the project's sdk 
                using Sdk attribute (from .NET Core Sdk 1.0.0-preview4) -->
   <Import Project="$(LanguageTargets)"/>

--- a/test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs
+++ b/test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs
@@ -472,6 +472,28 @@ namespace Microsoft.NET.Build.Tests
                 .Fail();
         }
 
+        [Fact]
+        public void Restore_succeeds_even_if_the_project_extension_is_for_a_different_language()
+        {
+            var testAsset = _testAssetsManager
+                .CopyTestAsset("AppWithLibrary")
+                .WithSource();
+
+            var libraryProjectDirectory = Path.Combine(testAsset.TestRoot, "TestLibrary");
+
+            var oldProjectFile = Path.Combine(libraryProjectDirectory, "TestLibrary.csproj");
+            var newProjectFile = Path.Combine(libraryProjectDirectory, "TestLibrary.fsproj");
+
+            File.Move(oldProjectFile, newProjectFile);
+
+            var restoreCommand = new RestoreCommand(Stage0MSBuild, libraryProjectDirectory, "TestLibrary.fsproj");
+
+            restoreCommand
+                .Execute()
+                .Should()
+                .Pass();
+        }
+
         [Theory]
         [InlineData(".NETStandard,Version=v1.0", new[] { "NETSTANDARD1_0" }, false)]
         [InlineData("netstandard1.3", new[] { "NETSTANDARD1_3" }, false)]

--- a/test/Microsoft.NET.TestFramework/Commands/RestoreCommand.cs
+++ b/test/Microsoft.NET.TestFramework/Commands/RestoreCommand.cs
@@ -14,8 +14,8 @@ namespace Microsoft.NET.TestFramework.Commands
         private List<string> _sources = new List<string>();
         private bool _captureStdOut;
 
-        public RestoreCommand(MSBuildTest msbuild, string projectPath)
-            : base(msbuild, projectPath)
+        public RestoreCommand(MSBuildTest msbuild, string projectPath, string relativePathToProject = null)
+            : base(msbuild, projectPath, relativePathToProject)
         {
         }
 

